### PR TITLE
Fix `EmrCreateJobFlowOperator` using deferrable mode with `wait_for_completion`

### LINKED ISSUE #40966: EmrCreateJobFlowOperator should not wait_for_completion if wait_for_completion=false deferrable=true
### Apache Airflow version

2.9.3

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

When setting global 
```
AIRFLOW__OPERATORS__DEFAULT_DEFERRABLE =true
EmrCreateJobFlowOperator(wait_for_completion=false)
```
The operator should adds a deferral trigger to wait for Jobflow create completion status.

### What you think should happen instead?

Since we did not ask to wait for completion, the deferrable argument should do nothing. The deferrable argument should only start a trigger when `wait_for_completion=true`

### How to reproduce

See deferrable code here:
https://github.com/apache/airflow/blob/460d2095cda6674dda4860d605573b8f1cec7de5/airflow/providers/amazon/aws/operators/emr.py#L827-L849

Sample code
```
EmrCreateJobFlowOperator(
        task_id="cluster_m5",
        wait_for_completion=false,
        deferrable=true
)
```

### Operating System

Mac OS 14.5 (23F79)

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by laksh-krishna-sharma**
---

### Summary of Changes
This PR addresses issue #40966 by modifying the logic in the `EmrCreateJobFlowOperator` to ensure that the `deferrable` argument only activates the deferral trigger when `wait_for_completion` is also set to `True`. Previously, the deferral trigger would be added regardless of the `wait_for_completion` setting, leading to unintended behavior.

### Changes Made:
- Updated the condition for deferring to check both `deferrable` and `wait_for_completion`.
- Adjusted the condition for synchronous waiting to occur only when `wait_for_completion=True` and `deferrable=False`.

Please review the changes, and let me know if further adjustments are needed.

Closes: https://github.com/apache/airflow/issues/40966

### CODE REVIEW

File: airflow/providers/amazon/aws/operators/emr.py [hunk 1]
```diff
@@ -824,7 +824,7 @@ def execute(self, context: Context) -> str | None:
                 job_flow_id=self._job_flow_id,
                 log_uri=get_log_uri(emr_client=self._emr_hook.conn, job_flow_id=self._job_flow_id),
             )
-        if self.deferrable:
+        if self.deferrable and self.wait_for_completion:
```
Comment (shahar1) [inline airflow/providers/amazon/aws/operators/emr.py]: 1. I'd consider refactoring both conditions as follows (a bit more readable IMO):
```python
if self.wait_for_completion:
    if self.deferrable:
       ...
    else:
       ...

``` 

2. Unit tests + docstring update for this behavior would be helpful.

File: airflow/providers/amazon/aws/operators/emr.py [hunk 2]
```diff
@@ -833,11 +833,9 @@ def execute(self, context: Context) -> str | None:
                     waiter_max_attempts=self.waiter_max_attempts,
                 ),
                 method_name="execute_complete",
-                # timeout is set to ensure that if a trigger dies, the timeout does not restart
-                # 60 seconds is added to allow the trigger to exit gracefully (i.e. yield TriggerEvent)
```
Comment (shahar1) [inline airflow/providers/amazon/aws/operators/emr.py:837]: Why did you delete these comments?

**Review by shahar1 — COMMENTED** [review]
Welcome to Apache Airflow and congrats for your first PR!
Good start, I have some comments :)

**Comment by laksh-krishna-sharma** [comment]
Thank you for letting me know, @shahar1. I will make the changes and add comments again . I'll update you once the problem is resolved."

**Comment by laksh-krishna-sharma** [comment]
@shahar1 sir please review the update.

File: airflow/providers/amazon/aws/operators/emr.py (hunk 1 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/operators/emr.py]: Is not it the case for 1? But I definitely agree on 2 :)

**Comment by vincbeck** [comment]
Can you please add unit tests?

**Comment by laksh-krishna-sharma** [comment]
@shahar1 sir could you please advise on the appropriate folder for placing unit test cases?

**Comment by shahar1** [comment]
> @shahar1 sir could you please advise on the appropriate folder for placing unit test cases?

`tests/providers/amazon/aws/operators/`, and there check all the files that start with `emr_`.

**Comment by laksh-krishna-sharma** [comment]
@shahar1 sir and @vincbeck sir I've added unit test cases and done my best to cover the expected behaviors. As I'm a beginner, I'd greatly appreciate your feedback and any suggestions for improvement.

**Review by vincbeck — APPROVED** [review]
Good job!

**Comment by vincbeck** [comment]
> @shahar1 sir and @vincbeck sir I've added unit test cases and done my best to cover the expected behaviors. As I'm a beginner, I'd greatly appreciate your feedback and any suggestions for improvement.

Thanks for doing it! The tests are pretty good!

**Review by shahar1 — APPROVED** [review]

**Comment by shahar1** [comment]
Looks great!

**Comment by vincbeck** [comment]
Some tests are failing, could you check that please?

**Comment by vincbeck** [comment]
https://github.com/apache/airflow/actions/runs/10561464118/job/29259176171?pr=41561

**Comment by laksh-krishna-sharma** [comment]
Thank you for your review @vincbeck and @shahar1. I will investigate the issue and address the failing tests as soon as possible. I'll update once the problem is resolved.

**Comment by laksh-krishna-sharma** [comment]
@shahar1 sir and @vincbeck sir I've added unit test cases and done my best to cover the expected behaviors. As I'm a beginner, I'd greatly appreciate your feedback and any suggestions for improvement but i am stuck in resolving these errors please sir can you guide me to resolve these issue

**Comment by shahar1** [comment]
> @shahar1 sir and @vincbeck sir I've added unit test cases and done my best to cover the expected behaviors. As I'm a beginner, I'd greatly appreciate your feedback and any suggestions for improvement but i am stuck in resolving these errors please sir can you guide me to resolve these issue

Hello, here's some guidance:
1. In `test_create_job_flow_deferrable` you need to set `self.operator.wait_for_completion = True` (otherwise it won't pass the first if).
1. In `test_deferrable_and_wait_for_completion`, when you call the `assert` statement you create a new instance of `EmrCreateJobFlowTrigger` - which will always be different than the instance that is created by the original function.
I'd suggest to mock `EmrCreateJobFlowTrigger` (let's call it `mock_trigger`) and then you could assert on `trigger=mock_trigger.return_value`. 
You could also create a separate assertion for the attributes `mock_trigger` to ensure that all the parameters are as expected.

Please let me know if you encounter any issues.

**Comment by laksh-krishna-sharma** [comment]
Thank you for your help, @shahar1 sir. I’ve resolved the issue, and all checks have now passed on GitHub. I really appreciate your guidance!

**Review by vincbeck — APPROVED** [review]

**Comment by vincbeck** [comment]
There is a conflict now :)

**Comment by laksh-krishna-sharma** [comment]
Sir, I resolved the branch conflict. Please review the updates.

**Comment by laksh-krishna-sharma** [comment]
@shahar1 sir and @vincbeck sir I've tried to resolve conflict and done my best to cover the expected behaviors. As I'm a beginner, I'd greatly appreciate your feedback and any suggestions for improvement but i am stuck in resolving these errors please sir can you guide me to resolve these issue

**Comment by shahar1** [comment]
> @shahar1 sir and @vincbeck sir I've tried to resolve conflict and done my best to cover the expected behaviors. As I'm a beginner, I'd greatly appreciate your feedback and any suggestions for improvement but i am stuck in resolving these errors please sir can you guide me to resolve these issue

Take a look at this part:
```python
        mock_defer.assert_called_once_with(
            trigger=EmrCreateJobFlowTrigger( # <- New instance
                job_flow_id=JOB_FLOW_ID,
                aws_conn_id=self.operator.aws_conn_id,
                waiter_delay=self.operator.waiter_delay,
                waiter_max_attempts=self.operator.waiter_max_attempts,
            ),
            method_name="execute_complete",
            timeout=timedelta(seconds=self.operator.waiter_max_attempts * self.operator.waiter_delay + 60),
        )
```

As I previously wrote, here you create a new instance of `EmrCreateJobFlowTrigger` - which means that it will always be different than the instance created within the [trigger](https://github.com/apache/airflow/blob/3cea4fdc9e5ff55c0fb87d3f57e0be3fba520f28/airflow/providers/amazon/aws/operators/emr.py#L833) attribute of `defer()`. Try to find a way to patch the creation of this instance, so they'll become the same entity.

**Comment by bitomukesh** [comment]
<div>
<h2>Changelist by Bito</h2>
    <p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>Bug Fix</b> - Fix EmrCreateJobFlowOperator Deferrable Mode Logic
</td>
<td width="70%">


<div>

-&nbsp;[`emr.py`](https://github.com/apache/airflow/pull/41561/files#diff-94612f7f88a5f95ca9e3f23051f39f1e0b6a4cadac1f533615db5480b379ed34) - Updated execute method to correctly handle deferrable mode with wait_for_completion
<div>

-&nbsp;[`test_emr_create_job_flow.py`](https://github.com/apache/airflow/pull/41561/files#diff-e34cc143955208a2aa5978741ef22ea3bfc1a4f222ffec8d497e797e7f5f3ef9) - Added and updated tests for EmrCreateJobFlowOperator's deferrable mode and wait_for_completion functionality
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

**Comment by laksh-krishna-sharma** [comment]
Thank you, @bitomukesh , for helping resolve the error! I really appreciate your assistance!"

**Comment by vincbeck** [comment]
Why closing it @laksh-krishna-sharma?
